File: android/Book-AndroidProgramming/CriminalIntentJ/app/src/main/java/kr/ac/ajou/heidi/criminalintentj/model/CrimeLab.java
package kr.ac.ajou.heidi.criminalintentj.model;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.os.Environment;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import kr.ac.ajou.heidi.criminalintentj.database.CrimeBaseHelper;
import kr.ac.ajou.heidi.criminalintentj.database.CrimeCursorWrapper;
import kr.ac.ajou.heidi.criminalintentj.database.CrimeDbSchema;

public class CrimeLab {
    private static CrimeLab sCrimeLab;

    private Context mContext;
    private SQLiteDatabase mDatebase;

    public static CrimeLab get(Context context) {
        if (sCrimeLab == null) {
            sCrimeLab = new CrimeLab(context);
        }
        return sCrimeLab;
    }

    private CrimeLab(Context context) {
        mContext = context.getApplicationContext();
        mDatebase = new CrimeBaseHelper(mContext).getWritableDatabase();

    }

    public List<Crime> getCrimes() {
        List<Crime> crimes = new ArrayList<>();

        CrimeCursorWrapper cursor = queryCrimes(null, null);

        try {

            cursor.moveToFirst();
            while (!cursor.isAfterLast()) {
                crimes.add(cursor.getCrime());
                cursor.moveToNext();
            }
        } finally {
            cursor.close();
        }

        return crimes;
    }

    public Crime getCrime(UUID id) {

        CrimeCursorWrapper cursor = queryCrimes(
                CrimeDbSchema.CrimeTable.Cols.UUID + " = ?",
                new String[]{id.toString()}
        );

        try {
            if (cursor.getCount() == 0) {
                return null;
            }

            cursor.moveToFirst();
            return cursor.getCrime();
        } finally {
            cursor.close();
        }
    }

    public void addCrime(Crime c) {
        ContentValues values = getContentValues(c);
        mDatebase.insert(CrimeDbSchema.CrimeTable.NAME, null, values);
    }

    public void removeCrime(Crime crime) {

    }

    private static ContentValues getContentValues(Crime crime) {
        ContentValues values = new ContentValues();
        values.put(CrimeDbSchema.CrimeTable.Cols.UUID, crime.getmId().toString());
        values.put(CrimeDbSchema.CrimeTable.Cols.TITLE, crime.getmTitle());
        values.put(CrimeDbSchema.CrimeTable.Cols.DATE, crime.getmDate().getTime());
        values.put(CrimeDbSchema.CrimeTable.Cols.SOLVED, crime.ismSolved() ? 1 : 0);
        values.put(CrimeDbSchema.CrimeTable.Cols.SUSPECT, crime.getmSuspect());

        return values;
    }

    public void updateCrime(Crime crime) {
        String uuidString = crime.getmId().toString();
        ContentValues values = getContentValues(crime);

        mDatebase.update(CrimeDbSchema.CrimeTable.NAME, values,
                CrimeDbSchema.CrimeTable.Cols.UUID + " = ?",
                new String[]{uuidString});
    }

    private CrimeCursorWrapper queryCrimes(String whereClause, String[] whereArgs) {
        Cursor cursor = mDatebase.query(
                CrimeDbSchema.CrimeTable.NAME,
                null,
                whereClause,
                whereArgs,
                null,
                null,
                null
        );

        return new CrimeCursorWrapper(cursor);
    }

    public File getPhotoFile(Crime crime) {
        File externalFilesDir = mContext
                .getExternalFilesDir(Environment.DIRECTORY_PICTURES);

        if (externalFilesDir == null) {
            return null;
        }

        return new File(externalFilesDir, crime.getPhotoFilename());
    }
}

File: android/Book-AndroidProgramming/CriminalIntentK/app/src/main/java/kr/ac/ajou/heidi/criminalintentk/model/CrimeLab.kt
package kr.ac.ajou.heidi.criminalintentk.model

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import kr.ac.ajou.heidi.criminalintentk.database.CrimeBaseHelper
import kr.ac.ajou.heidi.criminalintentk.database.CrimeCursorWrapper
import kr.ac.ajou.heidi.criminalintentk.database.CrimeDbSchema
import java.util.*

class CrimeLab private constructor(val context: Context) {

    var database: SQLiteDatabase = CrimeBaseHelper(context).writableDatabase
    val crimes: ArrayList<Crime> = arrayListOf()
        get() {
            val cursor = queryCrimes(null, null)

            cursor.use {c ->
                c.moveToFirst()
                while (!c.isAfterLast) {
                    c.getCrime()?.let {
                        crimes.add(it)
                    }
                    c.moveToNext()
                }
            }

            return field
        }

    companion object {
        var isCrime = false
        var crimeLab: CrimeLab? = null
        fun get(context: Context): CrimeLab? {
            if (!isCrime) {
                crimeLab = CrimeLab(context)
                isCrime = true

            }
            return crimeLab
        }

        fun getContentValues(crime: Crime): ContentValues {
            val values = ContentValues()
            values.put(CrimeDbSchema.CrimeTable.Cols.UUID, crime.id.toString())
            values.put(CrimeDbSchema.CrimeTable.Cols.TITLE, crime.title)
            values.put(CrimeDbSchema.CrimeTable.Cols.DATE, crime.date.time)
            values.put(CrimeDbSchema.CrimeTable.Cols.SOLVED, if (crime.solved) 1 else 0)

            return values
        }
    }


    fun getCrime(id: UUID): Crime? {
        val cursor = queryCrimes(
                CrimeDbSchema.CrimeTable.Cols.UUID + " = ?",
                arrayOf(id.toString())
        )

        cursor.use { cursor ->
            if (cursor.count == 0) {
                return null
            }

            cursor.moveToFirst()
            return cursor.getCrime()
        }
    }

    fun addCrime(c: Crime) {
        val values = getContentValues(c)
        database.insert(CrimeDbSchema.CrimeTable.NAME, null, values)
    }

    fun updateCrime(crime: Crime) {
        val uuidString = crime.id.toString()
        val values = getContentValues(crime)

        database.update(CrimeDbSchema.CrimeTable.NAME, values, CrimeDbSchema.CrimeTable.Cols.UUID + " = ?",
                arrayOf(uuidString))
    }

    private fun queryCrimes(whereClause: String?, whereArgs: Array<String>?): CrimeCursorWrapper {
        val cursor = database.query(
                CrimeDbSchema.CrimeTable.NAME,
                null,
                whereClause,
                whereArgs,
                null,
                null,
                null
        )

        return CrimeCursorWrapper(cursor)
    }
}